workbench-core/datasets/src/wbcDataSetsAuthorizationPlugin.test.ts

```diff
@@ -75,17 +139,87 @@ describe('wbcDataSetsAuthorizationPlugin tests', () => {
       auditService: audit
     });
     plugin = new WbcDataSetsAuthorizationPlugin(authzService);
+
+    jest
+      .spyOn(DynamicAuthorizationService.prototype, 'createIdentityPermissions')
+      .mockImplementation(async (params) => {
+        if (params.identityPermissions[0].identityType === 'USER') {
+          return mockUserPermissionResponse;
+        } else if (params.identityPermissions[0].action === 'UPDATE') {
+          return mockReadWritePermissionsResponse;
+        }
+        return mockReadOnlyPermissionsResponse;
+      });
   });
 
   beforeEach(() => {
     expect.hasAssertions();
   });
 
   describe('addAccessPermission tests', () => {
-    it('throws a notimplemented exception', async () => {
-      await expect(plugin.addAccessPermission(accessPermission)).rejects.toThrow(
-        new Error('Method not implemented.')
-      );
+    it('throws a invalidPermission exception when identityType is not USER or GROUP', async () => {
+      await expect(
+        plugin.addAccessPermission({
+          dataSetId: dataSetId,
+          authenticatedUserId: userId,
+          roles: [],
+          permission: {
+            identityType: fakeData,
+            identity: groupId,
+            accessLevel: 'read-only'
+          }
+        })
+      ).rejects.toThrow(new InvalidPermissionError('IdentityType just be "GROUP" or "USER".'));
+      expect(authzService.createIdentityPermissions).not.toBeCalled();
+    });
+
+    it('throws a invalidPermission exception when accessLevel is not read-only or read-write', async () => {
+      await expect(
+        plugin.addAccessPermission({
+          dataSetId: dataSetId,
+          authenticatedUserId: userId,
+          roles: [],
+          permission: {
+            identityType: 'GROUP',
+            identity: groupId,
+            // @ts-ignore to test poor JS usage.
+            accessLevel: fakeData
+          }
+        })
+      ).rejects.toThrow(new InvalidPermissionError('Access Level must be "read-only" or "read-write".'));
+      expect(authzService.createIdentityPermissions).not.toBeCalled();
+    });
+
+    it('creates an UPDATE permission when read-write access is requested.', async () => {
+      await expect(plugin.addAccessPermission(readWriteAccessPermission)).resolves.toStrictEqual({
+        data: {
+          dataSetId: dataSetId,
+          permissions: [
+            {
+              identityType: 'GROUP',
+              identity: groupId,
+              accessLevel: 'read-write'
+            }
+          ]
+        }
+      } as PermissionsResponse);
```

Comment:
instead of using an `as` here you might want to declare a typed variable so the compiler can tell us if anything is wrong / missing and then pass that in to `data`

Reply:
I removed it, but `as` caused the compiler to spit out exactly the boneheaded mistake I was making which caused the test to fail!